src/uci_options.h

```diff
@@ -0,0 +1,159 @@
+#pragma once
+
+#include "parsing/input_parsing.h"
+
+#include "fmt/base.h"
+#include <cstdint>
+#include <functional>
+#include <optional>
+#include <string>
+#include <string_view>
+#include <variant>
+
+namespace ucioption {
+
+/* uci types for easier creation */
+using string = std::string;
+using check = bool;
+using spin = uint64_t;
+
+template<typename T>
+struct Storage {
+    T value;
+    const T defaultValue;
+    const std::function<void(const T&)> callback;
+};
+
+using Variant = std::variant<
+    Storage<std::string>,
+    Storage<bool>,
+    Storage<uint64_t>>;
+
+struct Limits {
+    uint64_t min;
+    uint64_t max;
+};
+
+struct UciOption {
+    const std::string_view name;
+    Variant variant;
+    const std::optional<Limits> limits {};
+};
+
+template<typename T>
+constexpr auto make(std::string_view name, T defaultVal, std::function<void(const T&)> cb)
+{
+    return UciOption {
+        .name = name,
+        .variant = Storage<T> { .value = defaultVal, .defaultValue = defaultVal, .callback = std::move(cb) },
+    };
+}
+
+template<typename T>
+constexpr auto make(std::string_view name, T defaultVal, Limits limits, std::function<void(const T&)> cb)
+{
+    return UciOption {
+        .name = name,
+        .variant = Storage<T> { .value = defaultVal, .defaultValue = defaultVal, .callback = std::move(cb) },
+        .limits = std::move(limits),
+    };
+}
+
+constexpr bool handleInput(UciOption& option, std::string_view input)
+{
+    return std::visit([input, name = option.name, limits = option.limits](auto&& arg) {
+        bool error = false;
+
+        using T = std::decay_t<decltype(arg)>;
+        if constexpr (std::is_same_v<T, Storage<bool>>) {
+            if (input == "true") {
+                arg.value = true;
+            } else if (input == "false") {
+                arg.value = false;
+            } else {
+                error = true;
+            }
+        } else if constexpr (std::is_same_v<T, Storage<std::string>>) {
+            arg.value = input;
+        } else if constexpr (std::is_same_v<T, Storage<uint64_t>>) {
+            const auto inputNum = parsing::to_number(input);
+            if (!inputNum.has_value())
+                return false;
```

Comment:
Possible nitpick: why is there a mix of returning false directly, and setting `error` to false before returning it with no handling? The former seems less prone to accidentally setting data in the error case.

Reply:
Good catch, I think this is leftovers that I forgot to remove. Let me remove it :+1: